backend/associations/views/election.py

```diff
@@ -1,92 +1,220 @@
+    def perform_destroy(self, instance: Election):
+        # Only allow deletion if the election has not started.
+        if instance.started:
+            raise ValidationError("The election cannot be deleted anymore.")
 
-        # Check if the new Ballot object is valid.
-        choice_names = [
-            c[0] for c in Choice.objects.filter(election=election).values_list("name")
-        ]
-        for choice in serializer.validated_data["choices"]:
-            if choice.name not in choice_names:
-                raise ValidationError("Invalid choices provided.")
+        instance.delete()
 
-        if len(serializer.validated_data["choices"]) > election.max_choices_per_ballot:
-            raise ValidationError("Too many choices.")
+    @action(detail=True, methods=("put",), permission_classes=(VotePermission,))
+    def vote(self, *args, **kwargs):
+        serializer = VoteSerializer(data=self.request.data)
+        serializer.is_valid(raise_exception=True)
+
+        election = self.get_object()
 
         # Check if the election is active.
         if not election.is_active:
             raise PermissionDenied("This election is not active.")
 
         # Check if the user is allowed to vote.
-        if request.user.id not in [
-            voter[0] for voter in election.registered_voters.values_list("id")
-        ]:
+        if not election.voters.filter(
+            user=self.request.user, status="PENDING"
+        ).exists():
             raise PermissionDenied("You are not allowed to vote.")
 
-        # Check if the user has already voted.
-        if request.user.id in [voter[0] for voter in election.voters.values_list("id")]:
-            raise PermissionDenied("You have already voted.")
+        # Check if all the choices belong to the election.
+        if any(
+            (
+                choice.election != election
+                for choice in serializer.validated_data["choices"]
+            )
+        ):
+            raise ValidationError("An invalid choice was provided.")
 
-        # Save the object.
-        serializer.save(election=election)
-        headers = self.get_success_headers(serializer.data)
+        # Check if at least one choice but not too many choices have been provided.
+        if len(serializer.validated_data["choices"]) > election.max_choices_per_ballot:
+            raise ValidationError("Too many choices were provided.")
+
+        if len(serializer.validated_data["choices"]) == 0:
+            raise ValidationError("No choice was provided.")
 
-        # Add the voter to the Election object.
-        election.voters.add(request.user)
+        # Update the voter state.
+        voter = election.voters.get(user=self.request.user, status="PENDING")
+        voter.status = "ONLINE_VOTE"
+        voter.save()
+
+        # Update the choices.
+        for choice in serializer.validated_data["choices"]:
+            choice.number_of_online_votes += 1
+            choice.save()
```

Comment:
Should be done in a single SQL transaction 
https://docs.djangoproject.com/fr/2.2/topics/db/transactions/